### **Data Types**

Data types in Python define the type of value stored in a variable and determine the operations that can be performed on that data.

Python has the following data types built-in by default:

| Category | Python Types |
|---|---|
| **Text Type** | str |
| **Numeric Types** | int, float, complex |
| **Sequence Types** | list, tuple, range |
| **Mapping Type** | dict |
| **Set Types** | set, frozenset |
| **Boolean Type** | bool |

### **1. Text Data Type**

Strings are used to store text data. A string is represented using the str class and can be created using single, double or triple quotes.

### **2. Numeric Data Types**

Numeric data types are used to store numeric values. It can be an integer, floating number or even a complex number. Python supports three main numeric types:

- **Integers:** value is represented by int class. It contains positive or negative whole numbers (without fractions or decimals).

- **Float:** value is represented by float class. It is a real number with a floating-point representation, specified by a decimal point.

- **Complex:** It is represented by a complex class. It stores numbers with real and imaginary parts.

### **3. Sequence Data Types**

A sequence is an ordered collection of items, which can be of similar or different data types. Elements in a sequence can be accessed using indexing.
  
- **Lists:** Lists are ordered and mutable collections used to store multiple items in a single variable. Elements in a list can be of different data types and are accessed using indexing.

- **Tuple:** Tuples are ordered and immutable collections used to store multiple items in a single variable. Once created, tuple elements cannot be modified and are accessed using indexing.

- **Range:** Python function used to generate a sequence of numbers. It is commonly used in loops and can define a starting point, ending point, and step. The ending value is not included in the sequence.

### **4. Dictionary Data Type**

Dictionaries are used to store data in key:value pairs. Each key in a dictionary must be unique and values are accessed using their keys with square brackets [] or get() method.

### **5. Set Data Type**

Sets are unordered and mutable collections used to store unique elements. Since sets are unordered, elements cannot be accessed using indexing.

### **6. Boolean Data Type**

Boolean data type represents one of two values: True or False. It is mainly used in conditions and comparisons and is represented by the bool class.

### **Getting the Data Type**

In [48]:
# STRING
print(type("Python"))

# INTEGER
print(type(34))

# FLOAT
print(type(29.8))

# COMPLEX
print(type(2 + 1j))

# LIST
print(type(["apple", "banana", "cherry"]))

# TUPLE
print(type(("apple", "banana", "cherry")))

# RANGE
print(type(range(10)))

# DICTIONARY
print(type({"name": "John", "age": 36}))

# SET
print(type({"apple", "banana", "cherry"}))

# FROZENSET
print(type(frozenset({"apple", "banana", "cherry"})))

# BOOLEAN
print(type(True))

<class 'str'>
<class 'int'>
<class 'float'>
<class 'complex'>
<class 'list'>
<class 'tuple'>
<class 'range'>
<class 'dict'>
<class 'set'>
<class 'frozenset'>
<class 'bool'>


### **Type Casting**

The process of converting a value from one data type to another. Python provides built-in functions such as *int()*, *float()*, *str()*, and *bool()* to perform type conversions. It is especially useful when taking input from users.

### **Converting to int**

In [49]:
# string to integer conversion

x = "30"        
y = int(x)

print(y)
print(type(y))

print("--------------")

# float to integer conversion
x = 45.9       
y = int(x)

print(y)
print(type(y))

30
<class 'int'>
--------------
45
<class 'int'>


`int()` does not round the number. It removes the decimal portion.

### **Converting to float**

In [50]:
# integer to float conversion
x = 30.6
y = float(x)

print(y)
print(type(y))

print("--------------")

# string to float conversion
x = "45.9"
y = float(x)

print(y)
print(type(y))

30.6
<class 'float'>
--------------
45.9
<class 'float'>


### **Converting to str**

In [51]:
# integer to string conversion
age = 26
age = str(age)

print(age)
print(type(age))

26
<class 'str'>


This is useful when combining numbers with text.

In [52]:
age = 25

print("I'm " + str(age)+ " years old.")

I'm 25 years old.


Without `str()`, it produces a `TypeError` because Python cannot directly concatenate a string and an integer.

### **Converting to bool**

In [53]:
print(bool(0))
print(bool(1))

False
True


Generally, values such as **0**, **""**, **None**, and **empty** collections are considered False.

In [54]:
print(bool(0))
print(bool(""))

print(bool(10))
print(bool("Hello"))

False
False
True
True


### **Converting str to list**

In [55]:
x = "Python"
y = list(x)

print(y)

['P', 'y', 't', 'h', 'o', 'n']


### **Converting list to tuple**

In [56]:
x = [1, 2, 3, 4, 5]
y = tuple(x)

print(x)
print(type(y))

[1, 2, 3, 4, 5]
<class 'tuple'>


### **Converting tuple to list**

In [57]:
x = (1, 2, 3, 4, 5)
y = list(x)

print(y)
print(type(y))

[1, 2, 3, 4, 5]
<class 'list'>


### **Multiple Conversion**

Integer → Float → String

In [58]:
num = 30              # Integer
num = float(num)

print(num)            # Float

num = str(num)

print(num)            # String
print(type(num))

30.0
30.0
<class 'str'>


### **Type Casting with User Input**

#### **Without typecasting**

In [59]:
num1 = input("Enter first price: ")
num2 = input("Enter second price: ")

print(num1 + num2)  # Concatenation of strings

5020


If user enter: 

first price: 50

second price: 20

The output will be:

5020

Because both values are strings

#### **With typecasting**

In [60]:
num1 = int(input("Enter first price: "))
num2 = int(input("Enter second price: "))

print(num1 + num2)

70


Now instead of concatenation, python sees: 50 + 20